aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -98,18 +103,44 @@ compute_result_type(at::ArrayRef<OperandInfo> operands,
 
 std::tuple<Device, ScalarType> TensorIterator::compute_common_type() {
   // See [Result type computation] in TensorIterator.h
+
+  // in case of in-place operations, result type will be defined and needs to
+  // be preserved. Not an additional predicate becaue doesn't participate in
+  // zero-dim floating point promotion.
+  auto output = compute_result_type(operands_,
```

Comment:
I don't think this special casing of output operands is necessary and it deviates from the NumPy behavior.  Output operands are handled in the `compute_types` check that the common dtype matches the output type.

The "common type" is intended to be the type in which the computation is performed. If we end up supporting additional casting modes then it may not always match the output type.

For example, consider `torch.sign(float_tensor, out=int_tensor)`. It's an error in PyTorch and by default in NumPy, but allowed (and correct) with unsafe casting mode. It's important that the operation is performed in the floating-point domain and *then* casted to the integer domain. Otherwise, values like `0.3` will be incorrectly rounded to zero before the sign() operation is applied.